tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorFlow.java

```diff
@@ -138,4 +150,83 @@ private TensorFlow() {}
       throw e;
     }
   }
+
+  // to keep them from getting GC'd
+  private static final Set<GradFunc> gradientFuncs =
+      Collections.newSetFromMap(new IdentityHashMap<>());
+
+  private static synchronized boolean hasGradient(String opType) {
+    try (PointerScope scope = new PointerScope()) {
+      NativeStatus status =
+          GradOpRegistry.Global().Lookup(opType, new GradFunc(new PointerPointer<>(1)));
+      return status.ok();
+    }
+  }
+
+  /**
+   * Register a custom gradient function for ops of {@code opType} type.
+   *
+   * <p>Note that this only works with graph gradients, and will eventually be deprecated in favor
+   * of unified gradient support once it is fully supported by tensorflow core.
+   *
+   * @param opType the type of op to register the gradient for. Should usually be an {@code OP_NAME}
+   *     field, i.e. {@link Add#OP_NAME}.
+   * @param gradient the gradient function to use
+   * @return {@code true} if the gradient was registered, {@code false} if there was already a
+   *     gradient registered for this op
+   */
+  public static synchronized boolean registerCustomGradient(
+      String opType, RawCustomGradient gradient) {
+    if (hasGradient(opType)) {
+      return false;
+    }
+    GradFunc g = RawCustomGradient.adapter(gradient);
+    GradOpRegistry.Global().Register(opType, g);
+    gradientFuncs.add(g);
+    return true;
+  }
+
+  /**
+   * Register a custom gradient function for ops of {@code opClass} type. The actual op type is
+   * detected from the class's {@link GeneratedOpMetadata} annotation. As such, it only works on
+   * generated op classes.
+   *
+   * @param opClass the class of op to register the gradient for.
+   * @param gradient the gradient function to use
+   * @return {@code true} if the gradient was registered, {@code false} if there was already a
+   *     gradient registered for this op
+   * @throws IllegalArgumentException if {@code opClass} does not have a {@link GeneratedOpMetadata}
+   *     field.
+   */
+  public static synchronized <T extends RawOpInputs<?>> boolean registerCustomGradient(
+      Class<T> opClass, CustomGradient<T> gradient) {
+    GeneratedOpInputsMetadata metadata = opClass.getAnnotation(GeneratedOpInputsMetadata.class);
```

Comment:
Because the gradient builder lambda gets a `Input` object.  I only need it for type inference, if there was a way to get the input type from the op type it would be fine, but there isn't.  You can get an op object from the inputs object but not visa versa (since that info isn't available in eager mode, and we will want this to support the new gradient tape eventually).

Agreed re custom ops, I didn't think about that.